File: Fork/app/src/main/java/com/eshed/fork/data/model/Recipe.java
package com.eshed.fork.data.model;

import java.util.List;

public class Recipe {
    private int recipeID;
    private String name;
    private String imageURL;
    private String contributor;
    private List<Direction> directions;
    private List<Ingredient> ingredients;
    private List<String> tags;
    private int parentRecipeID;

    public Recipe(int recipeID, String name, String imageURL, String contributor, List<Ingredient> ingredients, List<Direction> directions, List<String> tags) {
        this(recipeID, name, imageURL, contributor, ingredients, directions, tags, -1);
    }

    public Recipe(int recipeID, String name, String imageURL, String contributor, List<Ingredient> ingredients, List<Direction> directions, List<String> tags, int parentRecipeID) {
        this.recipeID = recipeID;
        this.name = name;
        this.imageURL = imageURL;
        this.contributor = contributor;
        this.ingredients = ingredients;
        this.directions = directions;
        this.tags = tags;
        this.parentRecipeID = parentRecipeID;
    }

    public int getRecipeID() {
        return recipeID;
    }

    public String getName() {
        return name;
    }

    public String getImageURL() {
        return imageURL;
    }

    public List<Direction> getDirections() {
        return directions;
    }

    public List<Ingredient> getIngredients() {
        return ingredients;
    }

    public List<String> getTags() {
        return tags;
    }

    public String getContributor() {
        return contributor;
    }

    public int getParentRecipeID() {
        return parentRecipeID;
    }

    public void setParentRecipeID(int parentRecipeID) {
        this.parentRecipeID = parentRecipeID;
    }
}

File: Fork/app/src/main/java/com/eshed/fork/Data/model/Recipe.kt
package com.eshed.fork.Data.model

import com.google.firebase.database.Exclude
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class Recipe
@JvmOverloads constructor(
    var recipeID: Int = -1,
    var name: String = "",
    var imageURL: String = "",
    var contributor: String = "",
    var ingredients: MutableList<Ingredient> = mutableListOf(),
    var directions: MutableList<Direction> = mutableListOf(),
    var tags: MutableList<String> = mutableListOf<String>(),
    var parentRecipeID: Int = -1
) {
    fun deepCopy(
        recipeID: Int,
        name: String,
        parentRecipeID: Int
    ): Recipe = copy(
            recipeID = recipeID,
            name = name,
            parentRecipeID = parentRecipeID,
            ingredients = ingredients.map { it.copy() }.toMutableList(),
            directions = directions.map { it.copy() }.toMutableList(),
            tags = tags.map {  it.capitalize() }.toMutableList()
        )

    @Exclude
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "recipeID" to recipeID,
            "name" to name,
            "imageURL" to imageURL,
            "contributor" to contributor,
            "ingredients" to ingredients,
            "directions" to directions,
            "tags" to tags,
            "parentRecipeID" to parentRecipeID

        )
    }
}